### PR TITLE
[3.0] Add connection pool

### PR DESCRIPTION
**Pull request by EarthChen**
## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/DefaultConnectionPool.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        idleCount.decrementAndGet();
+        return connection;
+    }
+
+    @Override
+    public void release(Connection connection) {
+        if (!all.contains(connection)) {
+            return;
+        }
+        if (connection.usedCount.decrementAndGet() > 0) {
```
Comment (guohao) [inline dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/DefaultConnectionPool.java:85]: race condion occurs when threads call `acquire` and `release` concurrently

Comment (guohao) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/DefaultConnectionPool.java:85]: Some sync mechanisms are required or close connection asyncly

Comment (guohao) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/DefaultConnectionPool.java:85]: Some sync mechanisms are required or close connection asyncly

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/MultiplexProtocolConnectionManager.java [hunk 2]
```diff
@@ -41,10 +40,16 @@ public Connection connect(URL url) {
...
-    public void forEachConnection(Consumer<Connection> connectionConsumer) {
-        protocols.values().forEach(p -> p.forEachConnection(connectionConsumer));
+    public ConnectionPool getConnectionPool(URL url) {
+        final ConnectionManager manager = protocols.computeIfAbsent(url.getProtocol(), this::createSingleProtocolConnectionManager);
+        return manager.getConnectionPool(url);
     }
 
+//    @Override
+//    public void forEachConnection(Consumer<Connection> connectionConsumer) {
+//        protocols.values().forEach(p -> p.forEachConnection(connectionConsumer));
```
Comment (guohao) [inline dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/MultiplexProtocolConnectionManager.java:50]: just delete

Comment (guohao) [inline dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/MultiplexProtocolConnectionManager.java:50]: just delete

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/MultiplexProtocolConnectionManager.java [hunk 3]
```diff
@@ -41,10 +40,16 @@ public Connection connect(URL url) {
     }
 
     @Override
-    public void forEachConnection(Consumer<Connection> connectionConsumer) {
-        protocols.values().forEach(p -> p.forEachConnection(connectionConsumer));
+    public ConnectionPool getConnectionPool(URL url) {
```
Comment (guohao) [inline dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/MultiplexProtocolConnectionManager.java:43]: `ConnectionPool` is similiar to `ConnectionManager` , can these merged into one ?

Comment (guohao) [inline dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/MultiplexProtocolConnectionManager.java:43]: `ConnectionPool` is similiar to `ConnectionManager` , can these merged into one ?

**Review by guohao — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleInvoker.java [hunk 4]
```diff
@@ -278,7 +287,14 @@ public boolean isAvailable() {
         if (!super.isAvailable()) {
             return false;
         }
-        return connection.isAvailable();
+        ConnectionPoolEntry poolEntry = connectionPool.acquire();
+        try {
+            connection = poolEntry.getConnection();
+            return connection.isAvailable();
+        } finally {
+            connectionPool.release(poolEntry);
```
Comment (guohao) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleInvoker.java:295]: Why release here?

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/connection/DefaultConnectionPool.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    private final URL url;
+
+    public DefaultConnectionPool(URL url) {
+        this.url = url;
+    }
+
+
+    @Override
+    public ConnectionPoolEntry acquire() {
+        DefaultConnectionPoolEntry connection = cache.poll();
```
Comment (guohao) [inline dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/connection/DefaultConnectionPool.java:68]: Thread not safe here

**Review by guohao — CHANGES_REQUESTED** [review]

**Review by AlbumenJ — APPROVED** [review]

**Pull request closed without merging**